workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,202 @@
+            new Dictionary<uint, HashSet<IReaderInternal>>();
+
+        private readonly SpatialOSBehaviourLibrary behaviourLibrary;
+
+        public SpatialOSBehaviourManager(GameObject gameObject, SpatialOSBehaviourLibrary library)
         {
-            // Setup caches
+            behaviourLibrary = library;
+
+            var spatialComponent = gameObject.GetComponent<SpatialOSComponent>();
+            entity = spatialComponent.Entity;
+            spatialId = spatialComponent.SpatialEntityId;
+
+            foreach (var behaviour in gameObject.GetComponents<MonoBehaviour>())
+            {
+                var readerIds = library.GetRequiredReaderComponentIds(behaviour.GetType());
+                foreach (var id in readerIds)
+                {
+                    if (!componentIdToReaderSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToReaderSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToReaderSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                var writerIds = library.GetRequiredWriterComponentIds(behaviour.GetType());
+                foreach (var id in writerIds)
+                {
+                    if (!componentIdToWriterSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToWriterSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToWriterSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                numUnsatisfiedReadersOrWriters[behaviour] = readerIds.Count + writerIds.Count;
+            }
         }
 
-        public List<ReaderWriterImpl> GetReadersWriters(uint componentId)
+
+        public HashSet<IReaderInternal> GetReadersWriters(uint componentId)
         {
-            return null;
+            return componentIdToCreatedReadersWriters[componentId];
         }
 
         public void EnableSpatialOSBehaviours()
         {
-            // inject + enable all spatialOSBehavioursToEnable
+            foreach (var behaviour in spatialOSBehavioursToEnable)
+            {
+                var dict = behaviourLibrary.InjectAllReadersWriters(behaviour, entity);
+                behaviourToCreatedReadersWriters[behaviour] = dict;
+                foreach (var idToReaderWriter in dict)
+                {
+                    var id = idToReaderWriter.Key;
+                    var reader = idToReaderWriter.Value;
+                    if (!componentIdToCreatedReadersWriters.ContainsKey(id))
+                    {
+                        componentIdToCreatedReadersWriters[id] = new HashSet<IReaderInternal>();
+                    }
+
+                    componentIdToCreatedReadersWriters[id].Add(reader);
+                }
+
+                behaviour.enabled = true;
```

Comment:
Wasn't the deal that we inject on all behaviours first, and then enable them all ?

Reply:
This was also asked by Sami when the SpatialOSBehaviourLibrary was written. My understanding is that either no or not for this first iteration. @JonasImprobable ?

Reply:
I see no reason not too, its a tiny change to fix this.

Reply:
I pinged @JonasImprobable on slack, waiting for reply before changing this but I agree it's easy to do.

Reply:
Turns out I completely misunderstood the question, and you're right, I will change this